cdragontoolbox/wad.py

```diff
@@ -93,18 +93,34 @@ def read_data(self, f):
         elif self.type == 3:
             return zstd_decompress(data)
         elif self.type == 4:
-            # Data is split into individual subchunks that are zstd compressed
-            return zstd_decompress(data)
+            # Data is split into individual subchunks that may be zstd compressed
+            if subchunkTOC is not None:
+                subchunkEntries = [struct.unpack('<IIQ', subchunkTOC[16*(subchunkIndex := self.first_subchunk_index+i):16*(subchunkIndex+1)]) for i in range(self.subchunk_count)]
+                output = bytes()
+                offset = 0
+                for subchunk in subchunkEntries:
+                    subchunkData = data[offset:offset+subchunk[0]]
+                    if subchunk[0] == subchunk[1]: # assume uncompressed
+                        output += subchunkData
+                    else:
+                        output += zstd_decompress(subchunkData)
+                    offset += subchunk[0]
+                return output
+            else:
+                try:
+                    return zstd_decompress(data)
+                except:
+                    return data # possibly wrong, but what else to do when the subchunkTOC isn't given and decompression fails?
```

Comment:
I also thought about it, and at least for our code, not throwing or at least catching any exceptions is probably a must.

Take the warwick file I mentioned as an example: In the patch they added the failing subchunked file, they also added the subchunktoc file. That means that the subchunktoc hash isn't resolved and extracting the file or even just running the `guess_extensions` function will throw, completely denying access to the wad file.

We could probably add heuristic `.subchunktoc` hash guessing code to ensure this doesn't happen, but catching it in case it does should still be done.

Reply:
The using `subchunk_toc` parameter could be a good idea.
* `subchunk_toc = something`: decompress using provided subchunk data, fail on error
* `subchunk_toc = None`: assume subchunk is unknown, return the original data on error (original proposal)
* `subchunk_toc = False`: there is no suchunk data at all; fail on error

Using `False` as default would be safer, and we could use `None` internally in our code.

Reply:
I should have checked your changes first. I didn't saw you used the exception approach. 👍